### PR TITLE
Device event validation

### PR DESCRIPTION
**Pull request by jh-bate**
@darinkrauss validation and tests updated based on docs.

### CODE REVIEW

File: data/types/base.go [hunk 1]
```diff
@@ -283,6 +283,18 @@ func (d Datum) ToStringArray(fieldName string, errs validate.ErrorProcessing) *[
 	return &stringArray
 }
 
+func (d Datum) ToMap(fieldName string, errs validate.ErrorProcessing) *map[string]interface{} {
```
Comment (darinkrauss) [inline data/types/base.go]: I don't think you need to return a pointer to a map here. You should be able to return nil for a map (or array or interface).

File: data/types/device/alarm.go [hunk 2]
```diff
@@ -15,10 +15,15 @@ func init() {
 
 type Alarm struct {
 	AlarmType *string `json:"alarmType" bson:"alarmType" valid:"devicealarmtype"`
+	Status    *string `json:"status,omitempty" bson:"status,omitempty" valid:"-"`
```
Comment (darinkrauss) [inline data/types/device/alarm.go]: Do we want to standardize on where the `omitempty` occurs in the tag?  With a quick find I see that the vast majority have it second except for a few validators where it is first (such as `valid:"omitempty,required"`).

File: data/types/device/prime.go [hunk 3]
```diff
@@ -21,11 +20,18 @@ type Prime struct {
...
-		Message:           "Must be greater than 0.0",
-		AllowedFloatRange: &types.AllowedFloatRange{LowerLimit: 0.0},
+		Message:           "Must be >= 0.0 and <= 3.0",
+		AllowedFloatRange: &types.AllowedFloatRange{LowerLimit: 0.0, UpperLimit: 3.0},
+	}
+
+	tubingVolumeField = types.FloatDatumField{
+		DatumField:        &types.DatumField{Name: "volume"},
+		Tag:               "devicevolume",
+		Message:           "Must be >= 0.0 and <= 100.0",
```
Comment (darinkrauss) [inline data/types/device/prime.go]: We should decide whether we want these characters in the strings (that will be escaped in JSON).  Just makes it harder to read the output.

File: data/types/device/alarm.go (hunk 2 above)
Comment (jh-bate) [inline data/types/device/alarm.go]: its a requirement for validators to have it first and the other tags it is the opposite

File: data/types/device/prime.go (hunk 3 above)
Comment (jh-bate) [inline data/types/device/prime.go]: yeah agreed - I was going to do a bulk update when we looked at all message we are returning to client of the service

**Comment by darinkrauss** [comment]
A couple comments, but otherwise looks good. :shipit:

File: data/types/base.go (hunk 1 above)
Comment (jh-bate) [inline data/types/base.go]: true - i have update the code to reflect this

**Pull request merged**